service/history/historyEngine.go

```diff
@@ -2254,20 +2255,93 @@ type updateWorkflowAction struct {
 	createDecision bool
 }
 
-func (e *historyEngineImpl) updateWorkflowExecutionWithAction(
+type workflowLockedContext struct {
+	context workflowExecutionContext
+	release releaseWorkflowExecutionFunc
+	err     error
+}
+
+type updateWorkflowActionFunc func(builder mutableState) (*updateWorkflowAction, error)
+
+func (e *historyEngineImpl) updateWorkflowExecutionWithActionAndContextReload(
 	ctx ctx.Context,
 	domainID string,
 	execution workflow.WorkflowExecution,
-	action func(builder mutableState) (*updateWorkflowAction, error),
+	action updateWorkflowActionFunc,
 ) (retError error) {
 
+	// for execution with runID, just update with action
+	if execution.GetRunId() != "" {
+		workflowContext := e.getWorkflowLockedContext(ctx, domainID, execution)
+		return e.updateWorkflowExecutionWithContextAndAction(workflowContext, action)
+	}
+
+	// for execution without runID,
+	// reload current run in case continueAsNew caused unexpected workflow already completed error
+	var release releaseWorkflowExecutionFunc
+	for attempt := 0; attempt < retryCountForContextReload; attempt++ {
+		context, release, err0 := e.historyCache.getOrCreateWorkflowExecution(ctx, domainID, execution)
+		if err0 != nil {
+			return err0
+		}
+		msBuilder, err1 := context.loadWorkflowExecution()
+		if err1 != nil {
+			release(err1)
+			return err1
+		}
+
+		if msBuilder.IsWorkflowExecutionRunning() {
+			workflowContext := &workflowLockedContext{
+				context, release, err0,
+			}
+			return e.updateWorkflowExecutionWithContextAndAction(workflowContext, action)
+		}
+
+		// workflow is completed because of continue as new, fetch latest run
```

Comment:
how do you know the workflow is completed because of continue as new without checking the close status?
the current workflow can be terminated, completed, failed, continue-as-newed, timeouted

Reply:
ok, updated.